library/Test/Hspec/Expectations/Json/Internal.hs

```diff
@@ -66,40 +67,55 @@ newtype Subset = Subset Value
-  -- concerns. There are no good options here, so we make some concessions:
-  --
-  -- 1. It's expected you don't subset differently in different elements of the
-  --    same list. If you have an assertion that needs this behavior, do it
-  --    manually, please
-  --
-  -- 2. It's expected that sorting will be done after pruning, if you intend to
-  --    match irrespective of extra keys or ordering (shouldMatchJson does this)
-  --
-  -- Therefore, we grab the first element from the Subset Array (if present) and
-  -- prune all elements of the Superset Array using it. This ensures that
-  -- different sorts or length in the Superset side are preserved, but we
-  -- are still able to prune *before* the sorting required for matching, which
-  -- is important.
-  --
-  -- Other options such as sort-before-prune, or pair-wise pruning (with align
-  -- or zip) all correctly handle some cases but not all. And most importantly,
-  -- the cases those options don't handle come out as confusing assertion
-  -- failures.
-  --
-  (Array a, Array b) -> Array $ case b V.!? 0 of
-    Nothing -> a
-    Just y -> (\x -> pruneJson (Superset x) (Subset y)) <$> a
-
-  (x, _) -> x
+pruneJson (Superset sup) (Subset sub) = go (sup, normalizeArraysForPrunning sub)
+ where
+  go = \case
+    (Object a, Object b) ->
+      Object $
+        KeyMap.intersectionWith (curry go) a b
+    -- Pruning elements in Arrays is *extremely* tricky in that it interacts with
+    -- both sorting and matching in what should be a function independent of those
+    -- concerns. There are no good options here, so we make some concessions:
+    --
+    -- 1. It's expected you don't subset differently in different elements of the
+    --    same list. If you have an assertion that needs this behavior, do it
+    --    manually, please
+    --
+    -- 2. It's expected that sorting will be done after pruning, if you intend to
+    --    match irrespective of extra keys or ordering (shouldMatchJson does this)
+    --
+    -- Therefore, we grab the first element from the Subset Array (if present) and
+    -- prune all elements of the Superset Array using it. This ensures that
+    -- different sorts or length in the Superset side are preserved, but we
+    -- are still able to prune *before* the sorting required for matching, which
+    -- is important.
+    --
+    -- Other options such as sort-before-prune, or pair-wise pruning (with align
+    -- or zip) all correctly handle some cases but not all. And most importantly,
+    -- the cases those options don't handle come out as confusing assertion
+    -- failures.
+    --
+    (Array a, Array b) -> Array $ case b V.!? 0 of
+      Nothing -> a
+      Just y -> go . (,y) <$> a
+    (x, _) -> x
+
+-- Our prunning algorithm does not work when elements have different shapes, ala [{a:1}, {b:1}]
+-- This normalizes arrays so that all elements have the same shape by inserting null values.
```

Comment:
> by inserting null values

Setting aside the question of if we're going to add this behavior now or later for a moment, once we're all done, then

```
treatNullsAsMissing <> subsetActualToExpected TheNewWay
```

and

```
subsetActualToExpected TheNewWay <> treatNullsAsMissing
```

Are different behaviors, right?

One of them (not sure which) will prune away all the nulls added by the other. Is that OK?

Reply:
I had a thought that the ordering of these `Normalizer`s might make a difference, but in practice I don't think it happens.

> One of them (not sure which) will have treatNullsAsMissing prune away all the nulls added by subsetActualToExpected. Is that OK?

`subsetActualToExpected` doesn't add any keys that were not already present though

Reply:
I still don't understand how running `treatNullsAsMissing` after `subsetActualToExpected` won't undo its effects.

Does `subsetActualToExpected` not turn `[{ a: 1 }, { b: 1 }]` into `[{ a: 1, b: null}, { a:null, b: 1 }]`?

Does `treatNullsAsMissing` not turn `[{ a: 1, b: null}, { a:null, b: 1 }]` into `[{ a: 1 }, { b: 1 }]` ?

Which of these am I still not conceptually understanding?

Reply:
`subsetActualToExpected` can't add any keys that weren't already present, it merely prunes the ones that are not in the json that you are pruning against. `subsetActualToExpected [{ a:1 }, { b:1 }] [{ a:1, b:null}, { a:null, b:1 }]` -> `[{ a:1 }, { b:1 }]`

`treatNullsAsMissing` does what you describe to make the equality check work regardless of if you use `null` or omit the field.

Reply:
> `subsetActualToExpected` can't add any keys that weren't already present

You said this before and I truly have no idea what you mean.

The docs say

> This normalizes arrays so that all elements have the same shape by inserting null values
> [{a:1}, {b:1}] -> [{a:1, b:null}, {a:null, b:1}]

How does "inserting null values" not add keys that weren't present before, this certainly seems to add a `b` key that wasn't there before to the first element and an `a` key that wasn't there before to the second.

> it merely prunes the ones that are not in the json that you are pruning against

I thought that's what it did _before_, the thing that didn't work.

I swear I thought we had reached and understanding and now I'm so lost again!

```
subsetActualToExpected [{ a:1 }, { b:1 }] [{ a:1, b:null}, { a:null, b:1 }]
  -> [{ a:1 }, { b:1 }]
```

This example is just... `const`? Maybe, can you show an example where:

- The response side has the condition that didn't work before and this PR makes work
- `subsetActualToExpected` with the updated approach is used
- Show what expected and actual (or superset/subset) *both* look like post normalization and before assertion

?

Reply:
> How does "inserting null values" not add keys that weren't present before, this certainly seems to add a b key that wasn't there before to the first element and an a key that wasn't there before to the second. 

This is only done on the `Subset` not the `Superset` and we use `intersect` on the two. This means that if it wasn't already in the `Superset` then it can't magically appear there after the `intersect`

Reply:
> The response side has the condition that didn't work before and this PR makes work
```
actual = [{ bar: 1}, { foo: 1}]
expected = [{foo: 1}, {bar: 1}] 
```
> Show what expected and actual (or superset/subset) both look like post normalization and before assertion
```
 old prune: [{},{foo: 1}] -- We removed `bar` because it wasn't in the first element of expected
 new prune: [{ "bar": 1 }, { "foo": 1 }] -- We retained both keys because we found both somewhere in expected
```

Reply:
I think the fact that there is so much confusion here (I am equally 🤯 ), maybe the simpler `expandHeterogenousArrays` is the better route to go.

Reply:
I went for it in 3cf05c23c71814cf274db8b1c5e7d369d933e6a7, LMKWYT

Reply:
> This is only done on the `Subset` not the `Superset` and we use `intersect` on the two. This means that if it wasn't already in the `Superset` then it can't magically appear there after the `intersect`

I think that's besides my point though, but I still don't really follow it, so I'm not sure.

I do think you are saying here you will insert some key(s) with a value of `null` in the `Subset` side, yes? So what I was saying is that then calling `treatNullsAsMissing` recursively removes any keys with values of `null` from both sides, yes? How does the latter not undo the former?

I don't know, maybe it's not worth me understanding this point.

> maybe the simpler expandHeterogenousArrays is the better route to go

Well if that's _not_ what you've (conceptually) implemented, that explains my confusion. I thought my explanation [here](https://github.com/freckle/hspec-expectations-json/pull/18#discussion_r1282397519) _was_ (conceptually) using an expand-heterogeneous step:

> **After/changed**:
>
> - Normalize Subset so all elements of the array have same keys (adding null) <-- HERE
> - Locate first element of corresponding array in Subset
> - Prune all keys of Superset not in that element

I was further encouraged by you saying "Yep!" to me asking, "Is that right?"

So, back in this comment, I was really just pre-agreeing with you saying,

> that would make the ordering of `Normalizers` matter though? Maybe that's okay?

Because calling `treatNullsAsMissing` after `expandHeterogenousArrays` would surely be a no-op, and since I thought I was describing _current_ functionality, I thought that situation was already present.

:upside_down_face: 

But at the end of the day, I think we can end up at that place where it's simpler to understand, has the order dependence, but that's fine, and we're all happy?

Reply:
> I don't know, maybe it's not worth me understanding this point.

I think the key insight might be that the add nulls thing was just an implementation detail of `pruneJson :: Superset -> Subset -> Value`, which was changed to use the union of all keys in the `Subset` array when pruning the `Superset` instead of just the first element's keys. It never actually added null keys to the returned `Value`.

Regardless, I think the final API that we cooked up is pretty nice. Thanks for working through this with me! 

I feel like there's probably some kind of "Design an API" blogpost buried in here somewhere.